sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -168,6 +168,21 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
   }
 }
 
+/**
+ * Rewrite lateral joins by rewriting all dependent joins (if any) inside the right
+ * sub-tree of the lateral join and converting the lateral join into a base join type.
+ */
+object RewriteLateralJoin extends Rule[LogicalPlan] with PredicateHelper {
+
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case j @ Join(left, right, LateralJoin(joinType), condition, _) =>
+      val conditions = condition.map(splitConjunctivePredicates).getOrElse(Nil)
+      val newRight = DecorrelateInnerQuery.rewriteDomainJoins(left, right, conditions)
+      // TODO: handle the COUNT bug
```

Comment:
Can you add the jira number (SPARK-15370) here?

Reply:
Created a new ticket for handling the COUNT bug in lateral subqueries: https://issues.apache.org/jira/browse/SPARK-35551